allennlp/models/reading_comprehension/qanet.py

```diff
@@ -0,0 +1,262 @@
+
+        self._text_field_embedder = text_field_embedder
+
+        self._embedding_proj_layer = torch.nn.Linear(text_embed_dim, encoding_in_dim)
+        self._highway_layer = Highway(encoding_in_dim, num_highway_layers)
+
+        self._encoding_proj_layer = torch.nn.Linear(encoding_in_dim, encoding_in_dim)
+        self._phrase_layer = phrase_layer
+
+        self._matrix_attention = matrix_attention_layer
+
+        self._modeling_proj_layer = torch.nn.Linear(encoding_out_dim * 4, modeling_in_dim)
+        self._modeling_layer = modeling_layer
+
+        self._span_start_predictor = torch.nn.Linear(modeling_out_dim * 2, 1)
+        self._span_end_predictor = torch.nn.Linear(modeling_out_dim * 2, 1)
+
+        self._span_start_accuracy = CategoricalAccuracy()
+        self._span_end_accuracy = CategoricalAccuracy()
+        self._span_accuracy = BooleanAccuracy()
+        self._metrics = SquadEmAndF1()
+        self._dropout = torch.nn.Dropout(p=dropout_prob) if dropout_prob > 0 else lambda x: x
+
+        initializer(self)
+
+    def forward(self,  # type: ignore
+                question: Dict[str, torch.LongTensor],
+                passage: Dict[str, torch.LongTensor],
+                span_start: torch.IntTensor = None,
+                span_end: torch.IntTensor = None,
+                metadata: List[Dict[str, Any]] = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        """
+        Parameters
+        ----------
+        question : Dict[str, torch.LongTensor]
+            From a ``TextField``.
+        passage : Dict[str, torch.LongTensor]
+            From a ``TextField``.  The model assumes that this passage contains the answer to the
+            question, and predicts the beginning and ending positions of the answer within the
+            passage.
+        span_start : ``torch.IntTensor``, optional
+            From an ``IndexField``.  This is one of the things we are trying to predict - the
+            beginning position of the answer with the passage.  This is an `inclusive` token index.
+            If this is given, we will compute a loss that gets included in the output dictionary.
+        span_end : ``torch.IntTensor``, optional
+            From an ``IndexField``.  This is one of the things we are trying to predict - the
+            ending position of the answer with the passage.  This is an `inclusive` token index.
+            If this is given, we will compute a loss that gets included in the output dictionary.
+        metadata : ``List[Dict[str, Any]]``, optional
+            If present, this should contain the question ID, original passage text, and token
+            offsets into the passage for each instance in the batch.  We use this for computing
+            official metrics using the official SQuAD evaluation script.  The length of this list
+            should be the batch size, and each dictionary should have the keys ``id``,
```

Comment:
Is this still accurate?  I'm pretty sure there's no "id" key being used anywhere here anymore.

Reply:
Yes, this is outdated, but it's the same for the bidaf model. Should I change both of them?

Reply:
That would be great.